source/extensions/common/async_files/async_file_manager_thread_pool.cc

```diff
@@ -60,46 +50,121 @@ std::string AsyncFileManagerThreadPool::describe() const {
+  auto cancel_func = [state = entry.state_]() { state->store(QueuedAction::State::Cancelled); };
+  absl::MutexLock lock(&queue_mutex_);
+  queue_.push(std::move(entry));
   return cancel_func;
 }
 
+void AsyncFileManagerThreadPool::postCancelledActionForCleanup(
+    std::unique_ptr<AsyncFileAction> action) {
+  absl::MutexLock lock(&queue_mutex_);
+  cleanup_queue_.push(std::move(action));
+}
+
+void AsyncFileManagerThreadPool::executeAction(QueuedAction&& queued_action) {
+  using State = QueuedAction::State;
+  State expected = State::Queued;
+  std::shared_ptr<std::atomic<State>> state = std::move(queued_action.state_);
+  std::unique_ptr<AsyncFileAction> action = std::move(queued_action.action_);
+  if (!state->compare_exchange_strong(expected, State::Executing)) {
+    ASSERT(expected == State::Cancelled);
+    if (action->executesEvenIfCancelled()) {
+      action->execute();
+    }
+    return;
+  }
+  action->execute();
+  expected = State::Executing;
+  if (!state->compare_exchange_strong(expected, State::InCallback)) {
+    ASSERT(expected == State::Cancelled);
+    action->onCancelledBeforeCallback();
+    return;
+  }
+  if (queued_action.dispatcher_ == nullptr) {
+    // No need to bother arranging the callback, because a dispatcher was not provided.
+    return;
+  }
+  // If it is necessary to explicitly undo an action on cancel then the lambda will need a
+  // pointer to this manager that is guaranteed to outlive the lambda, in order to be able
+  // to perform that cancel operation on a thread belonging to the file manager.
+  // So capture a shared_ptr if necessary, but, to avoid unnecessary shared_ptr wrangling,
+  // leave it empty if the action doesn't have an associated cancel operation.
+  std::shared_ptr<AsyncFileManagerThreadPool> manager;
+  if (action->hasActionIfCancelledBeforeCallback()) {
+    manager = shared_from_this();
+  }
+  queued_action.dispatcher_->post([manager = std::move(manager), action = std::move(action),
+                                   state = std::move(state)]() mutable {
+    // This callback runs on the caller's thread.
+    State expected = State::InCallback;
+    if (state->compare_exchange_strong(expected, State::Done)) {
+      // Action was not cancelled; run the captured callback on the caller's thread.
+      action->onComplete();
+    } else {
+      ASSERT(expected == State::Cancelled);
+      if (manager == nullptr) {
+        // Action had a "do nothing" cancellation so we don't need to post a cleanup action.
+        return;
+      }
+      // If an action with side-effects was cancelled after being posted, its
+      // side-effects need to be undone as the caller can no longer receive the
+      // returned context. That undo action will need to be done on one of the
+      // file manager's threads, as it is file related, so post it to the thread pool.
+      manager->postCancelledActionForCleanup(std::move(action));
+      return;
```

Comment:
drop this return?

Reply:
Done, and outdented because the block was in an 'else' after a return too.